Snabb pflua integration

dpino opened the pull request:
- Fixed minor bug in _deps/Makefile_ (use bash as shell).
- Modified _filter.lua_ to use pflua instead of libpcap.
- Added a new packet filter app, _pflua_packet_filter_, using pflua. This app is based on the Snabbswitch's _packet_filter_ app.
- Added flag _-pflua_ to run Snabbswitch using pflua. The flag is needed to avoid including the pcap headers twice (ffi.cdata returns an error if the same C types are declared twice).
- Added some selftest examples to applications using filters. The selftests verify the filters used are working fine.

dpino committed:
Add new pflua packet filtering application

This is an implementation of the current packet filtering application
using pflua.

dpino committed:
Avoid loading pcap headers twice

dpino committed:
Adapt function filter

dpino committed:
Add selftest for ns_responder

dpino committed:
Add selftest for nd_light

dpino committed:
Use bash shell in Makefile

Necesarry to run 'cp pflua.git/{AUTHORS,COPYING,LICENSE}'.

dpino committed:
Refactor packet_filter benchmark

dpino commented:
There's a new *packet_filter* app that uses pflua. This app is called **pflua_packet_filter**. The app has a *selftest*, the same as the former packet filtering app (*apps/packet_filter*). To run the test:

```
snabbswitch/src $ sudo ./snabb -pflua -t apps.pflua_packet_filter.packet_filter
link report
6	sent on	packet_filter1.output -> sink1.input	(loss rate: 0%)
161	sent on	source1.output -> packet_filter1.input	(loss rate: 0%)
19	sent on	packet_filter2.output -> sink2.input	(loss rate: 0%)
43	sent on	source2.output -> packet_filter2.input	(loss rate: 0%)
v6.packets.tx: 6
v4.packets.tx: 19
selftest passed
```

I also ported a benchmark of the original packet filter app.

```
snabbswitch/src $ sudo ./snabb -pflua -t apps.pflua_packet_filter.packet_filter_benchmark
Run for 1 second ...
done
link report
161 sent on source.output -> repeater.input (loss rate: 0%))
6594560 sent on repeater.output -> packet_filter.input (loss rate: 0%))
245994 sent on packet_filter.output -> sink.input (loss rate: 0%))
```

While porting this benchmark I found a bug in pflua in the parsing of IPv6 addresses (https://github.com/Igalia/pflua/pull/96).

I modified the filter lib (*lib/pcap/filter.lua*), so it uses pflua. There are 3 apps that use the filter lib:
   * vpn/vpws.lua
   * ipv6/ns_responder.lua
   * ipv6/nd_light.lua

App *vpws* already included a *selftest*, the other two didn't. I added selftests for these apps, they just test the pflang expressions used in these apps. To run the tests:

**vpn/vws**

```
snabbswitch/src $ sudo ./snabb -pflua -t apps.vpn.vpws
link report
100	sent on	from_uplink.output -> vpntp.uplink	(loss rate: 0%)
100	sent on	vpntp.uplink -> to_uplink.input	(loss rate: 0%)
100	sent on	vpntp.customer -> to_customer.input	(loss rate: 0%)
100	sent on	from_customer.output -> vpntp.customer	(loss rate: 0%)
```

**ipv6/ns_responder**

```
snabbswitch/src $ sudo ./snabb -pflua -t apps.ipv6.ns_responder
link report
161	sent on	source1.output -> packet_filter.input	(loss rate: 0%)
9	sent on	packet_filter.output -> sink1.input	(loss rate: 0%)
selftest passed
```

**ipv6/nd_light**

```
snabbswitch/src $ sudo ./snabb -pflua -t apps.ipv6.nd_light
link report
161	sent on	source1.output -> packet_filter.input	(loss rate: 0%)
18	sent on	packet_filter.output -> sink1.input	(loss rate: 0%)
selftest passed
```

andywingo commented:
I think in the end we don't want separate pflua_packet_filter and packet_filter apps.  Given that you are introducing a -pflua flag -- not sure if that's a good idea either, but it's probably good for testing -- can you make the flag control the implementation of the existing packet_filter app, or are the configuration interfaces different?

andywingo commented:
The goal in the end would be to replace packet_filter with one based on pflua, of course, but we need something we can benchmark first.

dpino commented:
Both apps implement the same interface, so selecting one or another implementation depending on the -pflua flag won't be difficult to implement. I can work on that. 

OTOH, the reason why introducing the -pflua flag is to avoid the double inclusion of the libpcap headers. Pflua always includes these headers, but Snabbswitch does it too when the *pcap app* is used. Unfortunately, #ifdefs are not supported in LuaJIT (it's an open issue http://wiki.luajit.org/Open-Sponsorships#FFI-C-Pre-Processor) and including a header definition twice or more results into an error. So the best I could think of was to create this -pflua flag which what actually does is to avoid the inclusion of the pcap headers in Snabbswitch (as will be provided by pflua).

About benchmarking, the *packet filter app* includes a benchmark, so we will compare the app running with and without pflua.

dpino committed:
Packet filter pflua in the same dir as packet_filter

dpino commented:
I removed the *app/pflua_packet_filter* and provided a new PacketFilter class in the *app/packet_filter* directory. I split the packet_filter.lua file into two: packet_filter.lua (just provides the PacketFilter class) and packet_filter_test.lua (runs a test and picks different PacketFilter implementations depending on whether the -pflua flag is on or not). The *packet_filter_benchmark.lua* file does the same.

dpino commented:
I've been looking into the "bad" performance of pflua integrated in snabbswitch versus the current performance of the packet filtering app.

First of all, I made a slight modification (pending to update PR) that improves performance by a good margin (no longer orders of magnitude of 300% to 700% times slower, but below 100%). Secondly, I double checked the filter was not being compiled several times. It's only compiled once.

Then, I printed out the compiled Lua functions, both for the current app and pflua (the current app works like pflua, takes a filter expression, rather said configuration, and returns a Lua function that acts as a filter). I run the functions individually and checked both functions filter the same number of packets. This means both functions are semantically equivalent, apparently.

The compilation of a pflang filter by pflua has many aborted traces. To discard this could be a problem, I modified the code to load the function from a file, instead of compiling it in runtime. Loading the already compiled function from a file is faster. On the other hand, there are no performance penalties in Snabb's packet filtering app if the function is loaded from a file or compiled in runtime.

So after this I ended up with code that loads the filter (already compiled) from a file, either for the pflua case and for the current packet filtering app. And then I benchmarked both functions (execute 10 times, discard highest and lowest values). These are the results:

**Snabb's packet filtering app**

| **Step** | *MPPS*   |
|----------|----------|
| #1       | 39731200 |
| #2       | 28672000 |
| #3       | 75177984 |
| #4       | 75522048 |
| #5       | 40673280 |
| #6       | 77471744 |
| #7       | 40787968 |
| #8       | 77045760 |
| #9       | 77856768 |
| #10      | 78266368 |

AVG: 66,617.344 MPPS

**Packet filtering app with pflua**

| **Step** | *MPPS*   |
|----------|----------|
| #1       | 58286080 |
| #2       | 44916736 |
| #3       | 57982976 |
| #4       | 58056704 |
| #5       | 41508864 |
| #6       | 58630144 |
| #7       | 58630144 |
| #8       | 36962304 |
| #9       | 41320448 |
| #10      | 41230336 |

AVG: 54,861.824 MPPS

The current app is 20% faster than pflua.

Taking this into account I isolated both functions and execute them with ```pflua-match```. The pflua expression compiled is:

```
(icmp6 and src net 3ffe:501:0:1001::2/128 and dst net 3ffe:507:0:1:200:86ff:fe05:8000/116) or (ip6 and udp and src net 3ffe:500::/28 and dst net 3ffe:0501:4819::/64 and src portrange 2397-2399 and dst port 53)
```

This is the function evaluated by ```apps/packe_filter/packet_filter_benchmkar.lua```.

Here are the results:

**Function compiled by Snabb's packet filtering app**

| *Step* | *MPPS*    |
|--------|-----------|
| #1     | 93.984139 |
| #2     | 91.768137 |
| #3     | 91.157878 |
| #4     | 92.708215 |
| #5     | 93.632448 |
| #6     | 93.965557 |
| #7     | 94.513118 |
| #8     | 92.164841 |
| #9     | 92.197455 |
| #10    | 91.732004 |

AVG: 92.7823792 MPPS

**Function compiled by pflua**

| *Step* | *MPPS*    |
|--------|-----------|
| #1     | 76.240830 |
| #2     | 77.462574 |
| #3     | 76.017201 |
| #4     | 76.247829 |
| #5     | 76.573938 |
| #6     | 76.966456 |
| #7     | 77.321538 |
| #8     | 77.170282 |
| #9     | 76.664581 |
| #10    | 77.429569 |

AVG: 76.8094798 MPPS

The function compiled by Snabb's packet filtering app is also 20% faster.

In conclusion, there's nothing wrong with pflua only that the function compiled by snabb's packet filter app is more compact and more efficient. I read both functions and as for me they are semantically equivalent. Here's the code:

Expression:

```
(icmp6 and src net 3ffe:501:0:1001::2/128 and dst net 3ffe:507:0:1:200:86ff:fe05:8000/116) or (ip6 and udp and src net 3ffe:500::/28 and dst net 3ffe:0501:4819::/64 and src portrange 2397-2399 and dst port 53)
```

**Snabb's packet filtering app**

```lua
local ffi = require("ffi")
local bit = require("bit")
return function(buffer, size)
  repeat
    local ethertype = ffi.cast("uint16_t*", buffer + 12)
    if size < 54 then break end
    if ethertype[0] ~= 0xDD86 then break end
    local source_ip = ffi.cast("uint64_t*", buffer + 22)
    if source_ip[0] ~= 0x011000000105fe3fULL then break end
    if source_ip[1] ~= 0x0200000000000000ULL  then break end
    local dest_ip = ffi.cast("uint64_t*", buffer + 38)
    if dest_ip[0] ~= 0x010000000705fe3fULL then break end
    if 0x008005feff860002ULL ~= bit.band(0x00f0ffffffffffffULL, dest_ip[1]) then break end
    if buffer[20] ~= 58 then break end
    return true
  until false
  repeat
    local ethertype = ffi.cast("uint16_t*", buffer + 12)
    if size < 58 then break end
    if ethertype[0] ~= 0xDD86 then break end
    local source_ip = ffi.cast("uint64_t*", buffer + 22)
    if 0x000000000005fe3fULL ~= bit.band(0x00000000f0ffffffULL, source_ip[0]) then break end
    local dest_ip = ffi.cast("uint64_t*", buffer + 38)
    if dest_ip[0] ~= 0x000019480105fe3fULL then break end
    if buffer[20] ~= 17 then break end
    local source_port = buffer[54] * 0x100 + buffer[55]
    if source_port < 2397 or source_port > 2399 then break end
    local dest_port = ffi.cast("uint16_t*", buffer + 56)
    if dest_port[0] ~= 13568 then break end
    return true
  until false
  return false
end
```

**Pflua**

```lua
local floor=require('math').floor
local band=require('bit').band
local bxor=require('bit').bxor
local bor=require('bit').bor
local bswap=require('bit').bswap
local tobit=require('bit').tobit
local lshift=require('bit').lshift 
local rshift=require('bit').rshift
local bswap=require('bit').bswap
local cast=require('ffi').cast
    
return function(P,length)
   if length < 54 then return false end
   if cast("uint16_t*", P+12)[0] ~= 56710 then return false end
   local var2 = P[20]
   if var2 == 58 then goto L12 end
   do
      if length < 55 then return false end
      if var2 ~= 44 then goto L9 end
      if P[54] == 58 then goto L12 end
      goto L9
   end 
::L12::
   if cast("uint32_t*", P+22)[0] ~= 17169983 then goto L9 end
   do
      if cast("uint32_t*", P+26)[0] ~= 17825792 then goto L9 end
      if cast("uint32_t*", P+30)[0] ~= 0 then goto L9 end
      if cast("uint32_t*", P+34)[0] ~= 33554432 then goto L9 end
      if cast("uint32_t*", P+38)[0] ~= 117833279 then goto L9 end
      if cast("uint32_t*", P+42)[0] ~= 16777216 then goto L9 end
      if cast("uint32_t*", P+46)[0] ~= 4286971906 then goto L9 end
      if band(cast("uint32_t*", P+50)[0],15794175) == 8390142 then return true end
      goto L9
   end
::L9::
   if length < 58 then return false end
   if var2 == 17 then goto L40 end
   do
      if var2 ~= 44 then return false end
      if P[54] == 17 then goto L40 end
      return false
   end
::L40::
   if band(cast("uint32_t*", P+22)[0],-251658241) ~= 392767 then return false end
   if cast("uint32_t*", P+38)[0] ~= 17169983 then return false end
   if cast("uint32_t*", P+42)[0] ~= 6472 then return false end
   if var2 == 6 then goto L48 end
   do
      if var2 ~= 44 then goto L51 end
      do
         if P[54] == 6 then goto L48 end
         goto L51
      end
::L51::
      if var2 == 17 then goto L48 end
      if var2 ~= 44 then goto L57 end
      do
         if P[54] == 17 then goto L48 end
         goto L57
      end
::L57::
      if var2 == 132 then goto L48 end
      if var2 ~= 44 then return false end
      if P[54] == 132 then goto L48 end
      return false
   end
::L48::
   local var31 = rshift(bswap(cast("uint16_t*", P+54)[0]), 16)
   if var31 < 2397 then return false end
   if var31 > 2399 then return false end
   if var2 == 6 then goto L68 end
   do
      if var2 ~= 44 then goto L71 end
      do
         if P[54] == 6 then goto L68 end
         goto L71
      end
::L71::
      if var2 == 17 then goto L68 end
      if var2 ~= 44 then goto L77 end
      do
         if P[54] == 17 then goto L68 end
         goto L77
      end
::L77::
      if var2 == 132 then goto L68 end
      if var2 ~= 44 then return false end
      if P[54] == 132 then goto L68 end
      return false
   end
::L68::
   return cast("uint16_t*", P+56)[0] == 13568
end
```

This raises the question if we could generate Lua code that looks more like the one generated by Snabb's packet filter, for instance this is the code for checking a IPv6 address:

```lua
local source_ip = ffi.cast("uint64_t*", buffer + 22)
if source_ip[0] ~= 0x011000000105fe3fULL then break end
if source_ip[1] ~= 0x0200000000000000ULL  then break end
```
Versus the code we emit:

```lua
if cast("uint32_t*", P+22)[0] ~= 17169983 then goto L9 end
do
  if cast("uint32_t*", P+26)[0] ~= 17825792 then goto L9 end
  if cast("uint32_t*", P+30)[0] ~= 0 then goto L9 end
  if cast("uint32_t*", P+34)[0] ~= 33554432 then goto L9 end
  if cast("uint32_t*", P+38)[0] ~= 117833279 then goto L9 end
  if cast("uint32_t*", P+42)[0] ~= 16777216 then goto L9 end
  if cast("uint32_t*", P+46)[0] ~= 4286971906 then goto L9 end
  if band(cast("uint32_t*", P+50)[0],15794175) == 8390142 then return true end
  goto L9
end
```

I don't know for sure if the code by Snabb is better, it's just a gut feeling but I could verify it if that sounds interesting.

Lastly, please also notice that this is the best case for pflua, the already compiled filter is loaded from a file. If I compile the pflang expression in runtime, as it should likely be when pflua is integrated in snabb, the times are worse:

| **Step** | *MPPS*   |
|----------|----------|
| #1       | 30441472 |
| #2       | 48422912 |
| #3       | 34693120 |
| #4       | 31965184 |
| #5       | 35315712 |
| #6       | 49733632 |
| #7       | 35504128 |
| #8       | 49627136 |
| #9       | 49111040 |
| #10      | 31416320 |

AVG: 43,312.128

Snabb's packet filtering function is 53% times faster.

I also tried to compile the function using the BPF pipeline, and it gives worse results:

| **Step** | *MPPS*   |
|----------|----------|
| #1       | 28508160 |
| #2       | 36978688 |
| #3       | 39378944 |
| #4       | 31670272 |
| #5       | 34168832 |
| #6       | 24281088 |
| #7       | 38649856 |
| #8       | 39018496 |
| #9       | 26558464 |
| #10      | 40214528 |

AVG: 37,401.600

Snabb's packet filtering function is 78% times faster.

dpino commented:
Before updating the PR, would be possible to update pflua? The current PR contains a patch, already integrated in pflua, that fixes IPv6 parsing.

andywingo commented:
+@lukego because I know he enjoys optimization discussions

andywingo commented:
In the IP comparison case, you compare two 64-bit loads against pflua's 8 32-bit loads -- you miss a couple I think, because really that checks both destination and source IPv6 address, so there are a couple more 64-bit loads in the snabb code.  We could certainly turn adjacent 32-bit loads into 64-bit loads (or potentially 128 or 256 bit loads, with luajit help), if that's the issue.  I guess the first priority would be to know why the slowdown occurs when loading the function at runtime.  How long are these tests run?  Why are the traces aborting?  Is it something about the way we use setfenv?

andywingo commented:
I have updated the pflua checkout in the pflua branch.

dpino commented:
OK, thanks for noticing the issue with the source IP address, I will check that.

With regard to the other issues, I did some benchmarks. In neither of them there are issues related with interpreted code or excessive GC.

**Snabb's packet filter**

``` 
99%  Compiled
  -- 56%  filter-func.lua:pred
  -- 43%  pflua-match:filter
```

**pflua**

```
99%  Compiled
  -- 51%  pflua-match:filter
  -- 45%  filter-func-pflua.lua:pred
  --  3%  utils.lua:now
```
There are not NYI in pflua once the function is evaluated, the NYI happen when compiling the function. I filtered out "NYI: return to lower frame" as it has no effect in performance. Here's the trace:

```
3773:---- TRACE 10 abort parse.lua:309 -- NYI: bytecode 51
3779:---- TRACE 10 abort parse.lua:303 -- NYI: bytecode 51
3960:---- TRACE 12 abort app.lua:94 -- NYI: bytecode 72
8020:---- TRACE 33 abort expand.lua:981 -- NYI: bytecode 51
8513:---- TRACE 34 abort expand.lua:1028 -- NYI: bytecode 50
10437:---- TRACE 42 abort expand.lua:981 -- NYI: bytecode 51
10714:---- TRACE 43 abort expand.lua:1028 -- NYI: bytecode 50
10747:---- TRACE 43 abort expand.lua:1028 -- NYI: bytecode 50
11874:---- TRACE 45 abort expand.lua:981 -- NYI: bytecode 51
12387:---- TRACE 49 abort expand.lua:1030 -- NYI: bytecode 50
12582:---- TRACE 50 abort expand.lua:1030 -- NYI: bytecode 50
12695:---- TRACE 51 abort expand.lua:981 -- NYI: bytecode 51
12714:---- TRACE 51 abort expand.lua:1030 -- NYI: bytecode 50
12733:---- TRACE 51 abort expand.lua:1030 -- NYI: bytecode 50
12986:---- TRACE 54 abort expand.lua:981 -- NYI: bytecode 51
13431:---- TRACE 57 abort optimize.lua:136 -- NYI: bytecode 51
14060:---- TRACE 59 abort optimize.lua:128 -- NYI: bytecode 50
14072:---- TRACE 59 abort optimize.lua:128 -- NYI: bytecode 50
14736:---- TRACE 61 abort optimize.lua:136 -- NYI: bytecode 51
14763:---- TRACE 61 abort optimize.lua:136 -- NYI: bytecode 51
14822:---- TRACE 62 abort optimize.lua:136 -- NYI: bytecode 51
14849:---- TRACE 62 abort optimize.lua:136 -- NYI: bytecode 51
14880:---- TRACE 63 abort optimize.lua:128 -- NYI: bytecode 50
15460:---- TRACE 63 abort optimize.lua:136 -- NYI: bytecode 51
16725:---- TRACE 64 abort optimize.lua:136 -- NYI: bytecode 51
16752:---- TRACE 64 abort optimize.lua:136 -- NYI: bytecode 51
16779:---- TRACE 64 abort optimize.lua:136 -- NYI: bytecode 51
16806:---- TRACE 64 abort optimize.lua:136 -- NYI: bytecode 51
16833:---- TRACE 64 abort optimize.lua:136 -- NYI: bytecode 51
16871:---- TRACE 65 abort optimize.lua:128 -- NYI: bytecode 50
19370:---- TRACE 69 abort optimize.lua:136 -- NYI: bytecode 51
19698:---- TRACE 69 abort optimize.lua:136 -- NYI: bytecode 51
19751:---- TRACE 69 abort optimize.lua:174 -- NYI: bytecode 50
19804:---- TRACE 69 abort optimize.lua:174 -- NYI: bytecode 50
19857:---- TRACE 69 abort optimize.lua:174 -- NYI: bytecode 50
20409:---- TRACE 69 abort optimize.lua:136 -- NYI: bytecode 51
20595:---- TRACE 69 abort optimize.lua:136 -- NYI: bytecode 51
21058:---- TRACE 69 abort optimize.lua:136 -- NYI: bytecode 51
21111:---- TRACE 69 abort optimize.lua:174 -- NYI: bytecode 50
21164:---- TRACE 69 abort optimize.lua:174 -- NYI: bytecode 50
21231:---- TRACE 70 abort optimize.lua:174 -- NYI: bytecode 50
21261:---- TRACE 70 abort optimize.lua:174 -- NYI: bytecode 50
21314:---- TRACE 70 abort optimize.lua:174 -- NYI: bytecode 50
21343:---- TRACE 71 abort optimize.lua:128 -- NYI: bytecode 50
24801:---- TRACE 76 abort optimize.lua:136 -- NYI: bytecode 51
24819:---- TRACE 76 abort optimize.lua:136 -- NYI: bytecode 51
24837:---- TRACE 76 abort optimize.lua:136 -- NYI: bytecode 51
24855:---- TRACE 76 abort optimize.lua:136 -- NYI: bytecode 51
24874:---- TRACE 76 abort optimize.lua:128 -- NYI: bytecode 50
32600:---- TRACE 110 abort optimize.lua:136 -- NYI: bytecode 51
33414:---- TRACE 62 abort utils.lua:55 -- NYI: bytecode 72
33500:---- TRACE 63 abort utils.lua:55 -- NYI: bytecode 72
34864:---- TRACE 75 abort utils.lua:55 -- NYI: bytecode 72
35464:---- TRACE 80 abort optimize.lua:155 -- NYI: bytecode 50
35483:---- TRACE 80 abort optimize.lua:155 -- NYI: bytecode 50
35502:---- TRACE 80 abort optimize.lua:155 -- NYI: bytecode 50
35739:---- TRACE 85 abort optimize.lua:453 -- NYI: bytecode 51
36092:---- TRACE 118 abort optimize.lua:274 -- NYI: bytecode 51
36114:---- TRACE 118 abort optimize.lua:274 -- NYI: bytecode 51
36160:---- TRACE 118 abort optimize.lua:274 -- NYI: bytecode 51
36171:---- TRACE 118 abort optimize.lua:274 -- NYI: bytecode 51
36304:---- TRACE 119 abort optimize.lua:433 -- NYI: bytecode 72
36566:---- TRACE 121 abort optimize.lua:274 -- NYI: bytecode 51
36583:---- TRACE 121 abort optimize.lua:536 -- NYI: bytecode 50
36611:---- TRACE 121 abort optimize.lua:274 -- NYI: bytecode 51
36635:---- TRACE 121 abort optimize.lua:274 -- NYI: bytecode 51
36657:---- TRACE 121 abort optimize.lua:274 -- NYI: bytecode 51
36681:---- TRACE 121 abort optimize.lua:274 -- NYI: bytecode 51
36705:---- TRACE 121 abort optimize.lua:274 -- NYI: bytecode 51
36891:---- TRACE 123 abort optimize.lua:274 -- NYI: bytecode 51
36927:---- TRACE 123 abort optimize.lua:274 -- NYI: bytecode 51
36938:---- TRACE 123 abort optimize.lua:274 -- NYI: bytecode 51
36961:---- TRACE 123 abort optimize.lua:538 -- NYI: bytecode 50
37026:---- TRACE 124 abort optimize.lua:274 -- NYI: bytecode 51
37137:---- TRACE 126 abort optimize.lua:536 -- NYI: bytecode 50
37147:---- TRACE 126 abort optimize.lua:433 -- NYI: bytecode 72
37158:---- TRACE 126 abort optimize.lua:536 -- NYI: bytecode 50
37169:---- TRACE 126 abort optimize.lua:536 -- NYI: bytecode 50
37180:---- TRACE 126 abort optimize.lua:536 -- NYI: bytecode 50
37827:---- TRACE 129 abort optimize.lua:453 -- NYI: bytecode 51
37903:---- TRACE 130 abort optimize.lua:504 -- NYI: bytecode 50
37915:---- TRACE 130 abort optimize.lua:504 -- NYI: bytecode 50
38169:---- TRACE 132 abort optimize.lua:504 -- NYI: bytecode 50
38181:---- TRACE 132 abort optimize.lua:504 -- NYI: bytecode 50
38193:---- TRACE 132 abort optimize.lua:504 -- NYI: bytecode 50
38205:---- TRACE 132 abort optimize.lua:504 -- NYI: bytecode 50
38301:---- TRACE 134 abort optimize.lua:274 -- NYI: bytecode 51
38494:---- TRACE 136 abort optimize.lua:453 -- NYI: bytecode 51
38650:---- TRACE 137 abort optimize.lua:453 -- NYI: bytecode 51
38741:---- TRACE 137 abort optimize.lua:453 -- NYI: bytecode 51
38832:---- TRACE 137 abort optimize.lua:453 -- NYI: bytecode 51
39006:---- TRACE 140 abort optimize.lua:274 -- NYI: bytecode 51
39111:---- TRACE 141 abort optimize.lua:433 -- NYI: bytecode 72
39286:---- TRACE 142 abort optimize.lua:274 -- NYI: bytecode 51
39296:---- TRACE 142 abort optimize.lua:436 -- NYI: bytecode 72
39319:---- TRACE 142 abort optimize.lua:274 -- NYI: bytecode 51
39342:---- TRACE 142 abort optimize.lua:274 -- NYI: bytecode 51
39365:---- TRACE 142 abort optimize.lua:274 -- NYI: bytecode 51
39426:---- TRACE 144 abort optimize.lua:453 -- NYI: bytecode 51
39755:---- TRACE 147 abort optimize.lua:155 -- NYI: bytecode 50
39802:---- TRACE 148 abort optimize.lua:174 -- NYI: bytecode 50
39955:---- TRACE 149 abort optimize.lua:174 -- NYI: bytecode 50
39988:---- TRACE 149 abort optimize.lua:174 -- NYI: bytecode 50
40192:---- TRACE 151 abort optimize.lua:631 -- NYI: bytecode 51
40202:---- TRACE 151 abort optimize.lua:631 -- NYI: bytecode 51
40211:---- TRACE 151 abort optimize.lua:631 -- NYI: bytecode 51
40259:---- TRACE 151 abort optimize.lua:656 -- NYI: bytecode 50
40280:---- TRACE 151 abort optimize.lua:656 -- NYI: bytecode 50
40328:---- TRACE 151 abort optimize.lua:656 -- NYI: bytecode 50
40349:---- TRACE 151 abort optimize.lua:656 -- NYI: bytecode 50
40397:---- TRACE 151 abort optimize.lua:656 -- NYI: bytecode 50
40431:---- TRACE 153 abort optimize.lua:631 -- NYI: bytecode 51
40557:---- TRACE 154 abort optimize.lua:656 -- NYI: bytecode 50
40624:---- TRACE 154 abort optimize.lua:656 -- NYI: bytecode 50
40691:---- TRACE 154 abort optimize.lua:656 -- NYI: bytecode 50
40717:---- TRACE 154 abort optimize.lua:631 -- NYI: bytecode 51
41060:---- TRACE 157 abort optimize.lua:656 -- NYI: bytecode 50
41115:---- TRACE 157 abort optimize.lua:656 -- NYI: bytecode 50
41170:---- TRACE 157 abort optimize.lua:656 -- NYI: bytecode 50
41225:---- TRACE 157 abort optimize.lua:656 -- NYI: bytecode 50
41921:---- TRACE 162 abort optimize.lua:136 -- NYI: bytecode 51
41954:---- TRACE 162 abort optimize.lua:174 -- NYI: bytecode 50
42558:---- TRACE 167 abort utils.lua:55 -- NYI: bytecode 72
42692:---- TRACE 169 abort optimize.lua:128 -- NYI: bytecode 50
43277:---- TRACE 174 abort optimize.lua:274 -- NYI: bytecode 51
43323:---- TRACE 174 abort optimize.lua:274 -- NYI: bytecode 51
43384:---- TRACE 174 abort optimize.lua:274 -- NYI: bytecode 51
43445:---- TRACE 174 abort optimize.lua:274 -- NYI: bytecode 51
43783:---- TRACE 177 abort optimize.lua:274 -- NYI: bytecode 51
44510:---- TRACE 184 abort optimize.lua:274 -- NYI: bytecode 51
44562:---- TRACE 184 abort optimize.lua:274 -- NYI: bytecode 51
44614:---- TRACE 184 abort optimize.lua:274 -- NYI: bytecode 51
44666:---- TRACE 184 abort optimize.lua:274 -- NYI: bytecode 51
44959:---- TRACE 188 abort optimize.lua:274 -- NYI: bytecode 51
45229:---- TRACE 190 abort optimize.lua:274 -- NYI: bytecode 51
45265:---- TRACE 190 abort optimize.lua:274 -- NYI: bytecode 51
45485:---- TRACE 192 abort optimize.lua:274 -- NYI: bytecode 51
45639:---- TRACE 194 abort optimize.lua:274 -- NYI: bytecode 51
45675:---- TRACE 194 abort optimize.lua:274 -- NYI: bytecode 51
45711:---- TRACE 194 abort optimize.lua:274 -- NYI: bytecode 51
45757:---- TRACE 194 abort optimize.lua:274 -- NYI: bytecode 51
45778:---- TRACE 195 abort optimize.lua:631 -- NYI: bytecode 51
46457:---- TRACE 204 abort utils.lua:62 -- NYI: bytecode 72
46485:---- TRACE 204 abort utils.lua:62 -- NYI: bytecode 70
46511:---- TRACE 204 abort utils.lua:62 -- NYI: bytecode 72
46539:---- TRACE 204 abort utils.lua:62 -- NYI: bytecode 70
46566:---- TRACE 205 abort anf.lua:68 -- NYI: bytecode 51
46579:---- TRACE 205 abort anf.lua:68 -- NYI: bytecode 51
46585:---- TRACE 205 abort anf.lua:70 -- NYI: bytecode 51
46591:---- TRACE 205 abort anf.lua:68 -- NYI: bytecode 51
52954:---- TRACE 238 abort utils.lua:55 -- NYI: bytecode 72
53149:---- TRACE 239 abort utils.lua:55 -- NYI: bytecode 72
53252:---- TRACE 239 abort utils.lua:55 -- NYI: bytecode 72
53355:---- TRACE 239 abort utils.lua:55 -- NYI: bytecode 72
53458:---- TRACE 239 abort utils.lua:55 -- NYI: bytecode 72
53499:---- TRACE 240 abort utils.lua:55 -- NYI: bytecode 72
54375:---- TRACE 244 abort utils.lua:55 -- NYI: bytecode 72
56019:---- TRACE 248 abort utils.lua:55 -- NYI: bytecode 72
56047:---- TRACE 248 abort utils.lua:55 -- NYI: bytecode 72
59832:---- TRACE 259 abort backend.lua:148 -- NYI: bytecode 51
59838:---- TRACE 259 abort backend.lua:148 -- NYI: bytecode 51
60095:---- TRACE 261 abort backend.lua:148 -- NYI: bytecode 51
```
As a reference, here's the detail of the NYI bytecodes: UCLO (50), FNEW (51), ITERN (70), ISNEXT (72).

Instead of running pflua within Snabbswitch, it's possible to get a similar trace running pflua directly:
```
luajit -jdump=+rsT,/tmp/dump pflua-match v6.pcap \
"(icmp6 and src net 3ffe:501:0:1001::2/128 and dst net 3ffe:507:0:1:200:86ff:fe05:8000/116) or (ip6 and udp and src net 3ffe:500::/28 and dst net 3ffe:0501:4819::/64 and src portrange 2397-2399 and dst port 53)"
```

dpino commented:
Here's a summary of the aborted traces (due to NYI) per file:

**anf.lua**

```
anf.lua:22 -- NYI: bytecode 50
```

**backend.lua**

```
backend.lua:148 -- NYI: bytecode 51
backend.lua:174 -- NYI: bytecode 50
```

**bpf.lua**

**constants.lua**

**expand.lua**

```
expand.lua:1028 -- NYI: bytecode 50
expand.lua:1030 -- NYI: bytecode 50
expand.lua:981 -- NYI: bytecode 51
```

**libpcap.lua**

**optimize.lua**

```
optimize.lua:164 -- NYI: bytecode 50
optimize.lua:172 -- NYI: bytecode 51
optimize.lua:191 -- NYI: bytecode 50
optimize.lua:210 -- NYI: bytecode 50
optimize.lua:312 -- NYI: bytecode 51
optimize.lua:471 -- NYI: bytecode 72
optimize.lua:474 -- NYI: bytecode 72
optimize.lua:491 -- NYI: bytecode 51
optimize.lua:543 -- NYI: bytecode 50
optimize.lua:577 -- NYI: bytecode 50
optimize.lua:670 -- NYI: bytecode 51
optimize.lua:695 -- NYI: bytecode 50
optimize.lua:79 -- NYI: mixed sparse/dense table
```

**parse.lua**

**quickcheck.lua**

**savefile.lua**

**ssa.lua**

```
ssa.lua:90 -- NYI: bytecode 50
```

**types.lua**

**utils.lua**

```
utils.lua:55 -- NYI: bytecode 72
utils.lua:62 -- NYI: bytecode 70
utils.lua:62 -- NYI: bytecode 72
```

andywingo commented:
On Tue 03 Mar 2015 06:18, Diego Pino <notifications@github.com> writes:

> Here's a summary of the aborted traces (due to NYI) per file:
>
> anf.lua
>
> anf.lua:22 -- NYI: bytecode 50

etc

So these aren't important.  These aborts are made when compiling the
filter.  Is compiling the filter somehow taking a lot of the time?  If
so that needs fixing.  Otherwise I would guess that the benchmark is not
running for long enough.

Andy

dpino commented:
I measured how much time takes to compile the filter. Here are the results:

**Time to compile filter**

* Snabb: 0.00040400000000007 seg
* Pflua: 0.037258 seg

**Time to load function from file**

* Snabb: 0.00012699999999999 seg
* Pflua:   0.00015500000000002 seg

There's a considerable difference between compiling the filter and loading it from a file. I think that explains the difference in performance between both cases. The benchmark runs for 1 seg.

I increased the amount of time the benchmark runs. I measured the results every second, up to 5 seconds. Here is a line plot with the results.

![Snabb-vs-pflua](http://people.igalia.com/dpino/pflua/snabb-vs-pflua.png
"Snabb vs Pflua")

The difference between two measures is constant through time. If the main issue was the execesive time compiling the filter, the results would be better throughout time, being the 1 second benchmark the most penalized, but it's not the case.

Here is the file with the data used to generate the plot: [snabb-vs-pflua-data]: http://people.igalia.com/dpino/pflua/snabb-vs-pflua-data, and the csv file: [snabb-vs-pflua-data.csv]: http://people.igalia.com/dpino/pflua/snabb-vs-pflua-data.csv.

andywingo commented:
Do I understand that graph correctly that throughput increases over time, on a seconds timescale?  That is surprising to me.  Could it be instead that the Y access is "million packets processed" ?

Anyway, OK.  We've never attempted to optimize filter compilation.  Right now compiling one filter takes 37 milliseconds, from a cold start; OK.  I propose that we continue to ignore filter compilation time.  If it becomes an issue, we can try to optimize, but better to just fork off a process, compile and optimize there, and wait for the result to come back as text, and load that text.

So it would seem that the difference is in the filter function itself.  @dpino can you try manually replacing the uint32_t loads in the pflua filter with uint64 loads, using the snabb filter output as a guide?  Then we can see if that's the issue.  No need to modify pflua; we can just tweak the file on disk.

dpino commented:
The throughput in the chart is not normalized, I mean, is not divided by the number of seconds the benchmark was running, that's why it increases over time. What I wanted to see with that chart was whether both lines could converge at some point. This is the chart with the throughput divided by the number of seconds:

![Snabb-vs-pflua](http://people.igalia.com/dpino/pflua/snabb-vs-pflua-2.png
"Snabb vs Pflua")

I replaced uint32_t casts for uint64_t casts and throughput increased dramatically! Just 's/uint32_t/uint64_t', no more modifications. Here are the results:

**uint32_t**
```
Step 1: 76.906081
Step 2: 74.503802
Step 3: 75.142886
Step 4: 75.988306
Step 5: 76.251057
Step 6: 76.612333
Step 7: 72.802517
Step 8: 76.074271
Step 9: 75.669356
Step 10: 75.968172
```
AVG: 75.5918781

**uint64_t**
```
Step 1: 88.626225
Step 2: 87.208709
Step 3: 86.915045
Step 4: 86.401294
Step 5: 86.885012
Step 6: 83.938071
Step 7: 87.135045
Step 8: 83.782139
Step 9: 83.016830
Step 10: 87.672462
```
AVG: 86.1580832

It's still behind the filter compiled by Snabb which has an average performance of 91.641041 MPPS, but well, much better just with a minor change.

dpino commented:
For reference, this is the script I'm using to benchmark pflua with the function compiled by snabb and the function compiled by pflua.

```bash
#!/bin/bash

pflua=false

function exec_func {
    if [ $pflua == true ]; then
        ./pflua-match v6.pcap filter-func-pflua.lua
    else
        ./pflua-match v6.pcap filter-func.lua
    fi
}

if [ "$1" == "-pflua" ]; then
    pflua=true
fi

sum=0
for i in {1..10}; do
    val=`exec_func | grep -o '[0-9.]\+ MPPS' | grep -o '[0-9.]\+'`
    echo "Step $i: $val"
    sum=`echo "$sum + $val" | bc -l`
done

avg=`echo "scale=6; $sum / 10" | bc -l`
printf "\nAVG: $avg\n"
```

The functions ```filter-func.lua``` and ```filter-func-pflua.lua``` are available here: https://github.com/andywingo/snabbswitch/pull/3.

andywingo commented:
On Thu 05 Mar 2015 04:54, Diego Pino <notifications@github.com> writes:

> function exec_func {
>     if [ $pflua == true ]; then
>         ./pflua-match v6.pcap filter-func-pflua.lua
>     else
>         ./pflua-match v6.pcap filter-func.lua
>     fi
> }

This file is only 28 KB.  We're going to run into a lot of cache-related
things here, but I guess that is fine if we are comparing the two
directly.

Andy

dpino commented:
Sorry, this was wrong. The pflua function with uint32_t replaced as uint64_t  is returning 0 packets filtered, when it should return 6 packets filtered.

I modified the function, following Snabb's function logic and the performance doesn't change (75 MPPS approx). For the record, here's the function:

```lua
local floor=require('math').floor
local band=require('bit').band
local bxor=require('bit').bxor
local bor=require('bit').bor
local bswap=require('bit').bswap
local tobit=require('bit').tobit
local lshift=require('bit').lshift
local rshift=require('bit').rshift
local bswap=require('bit').bswap
local cast=require('ffi').cast

return function(P,length)
   if length < 54 then return false end
   if cast("uint16_t*", P+12)[0] ~= 56710 then return false end
   local var2 = P[20]
   local source_ip = cast("uint64_t*", P + 22)
   local dest_ip = cast("uint64_t*", P + 38)
   if var2 == 58 then goto L12 end
   do
      if length < 55 then return false end
      if var2 ~= 44 then goto L9 end
      if P[54] == 58 then goto L12 end
      goto L9
   end
::L12::
   do
      if source_ip[0] ~= 0x011000000105fe3fULL then goto L9 end
      if source_ip[1] ~= 0x0200000000000000ULL  then goto L9 end
      if dest_ip[0] ~= 0x010000000705fe3fULL then goto L9 end
      if 0x008005feff860002ULL ~= bit.band(0x00f0ffffffffffffULL, dest_ip[1]) then goto L9 end
      if P[20] ~= 58 then goto L9 end
      return true
   end
::L9::
   if length < 58 then return false end
   if var2 == 17 then goto L40 end
   do
      if var2 ~= 44 then return false end
      if P[54] == 17 then goto L40 end
      return false
   end
::L40::
   if 0x000000000005fe3fULL ~= bit.band(0x00000000f0ffffffULL, source_ip[0]) then return false end
   if dest_ip[0] ~= 0x000019480105fe3fULL then return false end
   if P[20] ~= 17 then return false end
   if var2 == 6 then goto L48 end
   do
      if var2 ~= 44 then goto L51 end
      do
         if P[54] == 6 then goto L48 end
         goto L51
      end
::L51::
      if var2 == 17 then goto L48 end
      if var2 ~= 44 then goto L57 end
      do
         if P[54] == 17 then goto L48 end
         goto L57
      end
::L57::
      if var2 == 132 then goto L48 end
      if var2 ~= 44 then return false end
      if P[54] == 132 then goto L48 end
      return false
   end
::L48::
   local var31 = rshift(bswap(cast("uint16_t*", P+54)[0]), 16)
   if var31 < 2397 then return false end
   if var31 > 2399 then return false end
   if var2 == 6 then goto L68 end
   do
      if var2 ~= 44 then goto L71 end
      do
         if P[54] == 6 then goto L68 end
         goto L71
      end
::L71::
      if var2 == 17 then goto L68 end
      if var2 ~= 44 then goto L77 end
      do
         if P[54] == 17 then goto L68 end
         goto L77
      end
::L77::
      if var2 == 132 then goto L68 end
      if var2 ~= 44 then return false end
      if P[54] == 132 then goto L68 end
      return false
   end
::L68::
   return cast("uint16_t*", P+56)[0] == 13568
end
```

dpino commented:
FWIW, there are unnecessary libraries included, but that doesn't have any effect in performance. Already tested.

lukego commented:
Just a quick and general note:

Snabb applications will typically be aiming for around 5-10 Mpps per core. The most interesting metric for features like packet filtering might be what fraction of the processing budget they will consume (rather than how many packets per second they can process in isolation).

For example, a packet filter processing 50 Mpps would consume an estimated 10% of the processing budget for a 5Mpps-per-core application where as 70 Mpps packet filter would consume more like 7%. Both cases sound pretty reasonable: the application developer's performance problems will probably be elsewhere.

lukego commented:
Corollaries:

1. Applications will want to measure the performance of new features in terms of their own end-to-end benchmarks. For example, Snabb NFV will care about the impact on benchmarks like "generate load through a VM" and running iperf between guests.
2. There will be diminishing returns on optimizing individual networking functions. (The slowest features will have the most influence on overall system performance.)
3. Predictable performance is really valuable. Application developers want to be able to say that "if you setup your application within these parameters then your minimum packet rate will be X."

Here is an example of the kind of performance data that networking manufacturers publish (comparative performance of many different Juniper firewalls): http://www.juniper.net/us/en/products-services/security/srx-series/. (More examples can be found for googling for "data sheets" of various products.)

This example does not inspire so much confidence because the actual limiting factor in a real network is likely to be something not shown e.g. packets-per-second (datasheet talks about "maximum Gbps" which could be assuming e.g. 9000 byte packets). Anecdotally, I have met network operators whose top-of-the-line Juniper SRX firewalls have been melting down under ordinary load much below the marketing numbers. That is always the concern for people operating networks: will it keep up with the rest of my network elements, and will that be true even if I modify the configuration.

lukego commented:
Last corollary :-)

Speaking with my application developer hat on, the main reason that I want pflua support is to allow users to configure more sophisticated packet filtering rules. If it also speeds up the application then that is a bonus. (The most important thing is not to massively slow it down when I least expect it :-))

dpino commented:
Considering Luke's last comments, I think the performance, although doesn't match the former packet filtering app, is now good enough.

I have pushed the latest changes and added a comment about implementing a cache for already compiled expressions.

Most likely there are going to be conflicts after rebasing against snabbswitch/master. I can rebase my commits. I think it'd a good idea also to update pflua to the latest version before sending the PR to snabbswitch/master.

andywingo commented:
Agreed.  dpino you can update pflua yourself -- just cd deps and make update-pflua, then commit.

dpino committed:
Update pflua

dpino commented:
@andywingo OK, done. However, it's necessary you rebase your snabbswitch fork with current snabbswitch/master. After that there are going to be conflicts, at least in these batch of patches (I've already tested it in a local branch).

dpino commented:
@andywingo  @lukego I moved this PR to a new repository https://github.com/dpino/snabbswitch/pull/1
